### DISCUSSION TITLE
foreign key constraint not enforced by sqlalchemy

### DISCUSSION THREAD

**Discussion by kerenduchan** [sqlalchemy/sqlalchemy#9111] [Usage Questions] [answered]
I wrote this sample code which defines a schema with two models, **Book** and **Author**. 
**Book.author_id** is a foreign key that references **Author.id**.
https://github.com/kerenduchan/sqlalchemy-foreign-key/blob/main/main.py

- Creating a book with a **nonexistent** author_id through sqlachemy succeeds, though it should fail.
- Running the query on the db fails with a foreign key constraint error, as expected.
  insert into books (title, author_id) values ("Book 2", 1)

Why doesn't sqlalchemy enforce the foreign key?

Thank you for your help.

**Comment by lelit** [comment]
That's not a thing SA controls, but rather a configurable option of `sqlite`: see https://www.sqlite.org/pragma.html#pragma_foreign_key_check

**Comment by lelit** [comment, marked as answer]
You can enable it, for example with something like:
```python
@event.listens_for(Engine, "connect")
def enable_sqlite_fks(dbapi_connection, connection_record):
    cursor = dbapi_connection.cursor()
    cursor.execute("PRAGMA foreign_keys=ON")
    cursor.close()
```

**Comment by kerenduchan** [comment]
Thank you! It worked.

This is the finished test code, for future reference:

```py
import asyncio
from sqlalchemy import event
from sqlalchemy.orm import sessionmaker, declarative_base
from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy import Column, Integer, String, ForeignKey


engine = create_async_engine('sqlite+aiosqlite:///library.db')


# add enforcement for foreign keys
@event.listens_for(engine.sync_engine, "connect")
def enable_sqlite_fks(dbapi_connection, connection_record):
    cursor = dbapi_connection.cursor()
    cursor.execute("PRAGMA foreign_keys=ON")
    cursor.close()


session_maker = sessionmaker(
    bind=engine,
    class_=AsyncSession,
    expire_on_commit=False,
    autocommit=False,
    autoflush=False,
)


Base = declarative_base()


class Author(Base):
    __tablename__ = "authors"
    id = Column(Integer, primary_key=True, autoincrement=True)
    name = Column(String, unique=True, nullable=False)


class Book(Base):
    __tablename__ = "books"
    id = Column(Integer, primary_key=True, autoincrement=True)
    title = Column(String, unique=True, nullable=False)
    author_id = Column(Integer, ForeignKey(Author.id), nullable=False)


async def init_db():
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.drop_all)
        await conn.run_sync(Base.metadata.create_all)


async def create_book():
    book = Book(title="Book 1", author_id=1)
    async with session_maker() as session:
        session.add(book)
        await session.commit()


async def main():
    await init_db()
    await create_book()

if __name__ == "__main__":
    asyncio.run(main())
```

**Comment by kerenduchan** [comment]
@lelit 
I have a follow-up question: 
Is this enforcement performed at the engine (db client)? 
Does that mean that the sqlite db does not contain the enforcement within itself? 
So someone can just write a client that doesn't have this foreign key enforcement, and insert dangling foreign keys into the db?
Can't I enforce the foreign keys at the db (server)?

Thank you.

**Reply by lelit** [reply]
With `sqlite` there isn't a _client_ and _server_ distinction as in other databases.
As explained in https://www.sqlite.org/foreignkeys.html#fk_enable, you necessarily must enable the check for each connection: so yes, an app that does not do that may be able to insert "wrong" data into a foreign key column.

**Comment by kerenduchan** [comment]
Thank you @lelit for the explanation! Always good to understand things better.
